ast/src/analyzed/mod.rs

```diff
@@ -621,12 +628,56 @@ pub enum SymbolKind {
 pub enum FunctionValueDefinition {
     Array(ArrayExpression<Reference>),
     Expression(TypedExpression),
-    TypeDeclaration(EnumDeclaration),
-    TypeConstructor(Arc<EnumDeclaration>, EnumVariant),
+    TypeDeclaration(TypeDeclaration),
+    TypeConstructor(TypeConstructor),
     TraitDeclaration(TraitDeclaration),
     TraitFunction(Arc<TraitDeclaration>, TraitFunction),
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub enum TypeDeclaration {
+    Enum(EnumDeclaration),
+    Struct(StructDeclaration),
+}
+
+impl Children<Expression> for TypeDeclaration {
+    fn children(&self) -> Box<dyn Iterator<Item = &Expression> + '_> {
+        match self {
+            TypeDeclaration::Enum(enum_decl) => enum_decl.children(),
+            TypeDeclaration::Struct(struct_decl) => struct_decl.children(),
+        }
+    }
+
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut Expression> + '_> {
+        match self {
+            TypeDeclaration::Enum(enum_decl) => enum_decl.children_mut(),
+            TypeDeclaration::Struct(struct_decl) => struct_decl.children_mut(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub enum TypeConstructor {
+    Enum(Arc<EnumDeclaration>, EnumVariant),
+    Struct(Arc<StructDeclaration>, (String, Type)),
```

Comment:
A type constructor for struct and a struct expression are the same thing. Is this the case here or do you make a distinction? I have the impression that we don't need struct type constructors but I might be wrong.

Reply:
I used it mainly in the Statement processor when processing the Struct declaration. `TypeDeclaration` declares the new type of the struct and `TypeConstructor`, constructed the types of the internal functions (similar to enums, as you mentioned), using the type declarations of each one of the functions inside the declaration. 

For example: `struct Dot { x: int, y: int }`
would give us a `TypeDeclaration` Dot and two `TypeConstructor` Dot::x and Dot::y.

On the other hand, the Struct expression is used at the time of instantiation, for example `let f: int -> Dot = |i| Dot{a: 0, b: i};`

is this wrong?

My biggest question here is: somehow I have to process the internal types so that they are available later, but if they are not TypeConstructor, what should they be?